### PR TITLE
init backgroundremover at 0.2.1

### PR DESCRIPTION
**Pull request by JayRovacsek**
###### Description of changes

This change set implements the backgroundremover python package and requisite dependencies that do not currently exist within nixpkgs.
Backgroundremover is a package to remove backgrounds from images and video and the homepage can be found [here](https://github.com/nadermx/backgroundremover)

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [X] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Result of `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"` run on x86-linux [1](https://github.com/Mic92/nixpkgs-review)

<details><summary>6 packages added:</summary><p>

- python3.10-backgroundremover
- python3.10-commandlines
- python3.10-hsh
- python3.10-PyMatting
- python3.11-commandlines
- python3.11-hsh

</p></details>

<details><summary>12 packages built:</summary><p>

- python310Packages.backgroundremover
- python310Packages.backgroundremover.dist
- python310Packages.commandlines
- python310Packages.commandlines.dist
- python310Packages.hsh
- python310Packages.hsh.dist
- python310Packages.pymatting
- python310Packages.pymatting.dist
- python311Packages.commandlines
- python311Packages.commandlines.dist
- python311Packages.hsh
- python311Packages.hsh.dist

</p></details>

Result of `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review) - note this is under remote builders

<details><summary>6 packages added:</summary><p>

- python3.10-backgroundremover
- python3.10-commandlines
- python3.10-hsh
- python3.10-PyMatting
- python3.11-commandlines
- python3.11-hsh

</p></details>

<details><summary>12 packages built:</summary><p>

- python310Packages.backgroundremover
- python310Packages.backgroundremover.dist
- python310Packages.commandlines
- python310Packages.commandlines.dist
- python310Packages.hsh
- python310Packages.hsh.dist
- python310Packages.pymatting
- python310Packages.pymatting.dist
- python311Packages.commandlines
- python311Packages.commandlines.dist
- python311Packages.hsh
- python311Packages.hsh.dist

</p></details>

Will add nixpkgs-review outputs from aarch64 and x86_64 darwin as soon as I can

### CODE REVIEW

**Comment by JayRovacsek** [comment]
Ah, just noticed; this closes https://github.com/NixOS/nixpkgs/issues/230522 :+1:

**Comment by AndersonTorres** [comment]
How dare you tell your contribution follows CONTRIBUTING guide?

**Comment by RaitoBezarius** [comment]
> How dare you tell your contribution follows CONTRIBUTING guide?
> 
> Have you ever read any code from this repo?

Please be more respectful of contributors, if you cannot afford to be nice, please do not review or comment.
This is needlessly aggressive.

**Review by JulienMalka — CHANGES_REQUESTED** [review]
Hello @JayRovacsek, thanks for your contribution.
You'll need to split it as 1 separate commit for each package added, and follow the proper commit naming convention that you can find in the contributing guide :)

File: pkgs/development/python-modules/backgroundremover/default.nix [hunk 1]
```diff
@@ -0,0 +1,53 @@
+{ buildPythonPackage, fetchPypi, lib, certifi, charset-normalizer, ffmpeg-python
```
Comment (lucasew) [inline pkgs/development/python-modules/backgroundremover/default.nix:1]: Write this parameters in this form:
```
{ parameter1
, parameter2
, parameter3
}:
```

File: pkgs/development/python-modules/backgroundremover/default.nix [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+, filelock, filetype, hsh, idna, more-itertools, moviepy, numpy, pillow
+, pymatting, pysocks, requests, scikitimage, scipy, six, torch, torchvision
+, tqdm, urllib3, waitress, gdown }:
+buildPythonPackage rec {
```
Comment (lucasew) [inline pkgs/development/python-modules/backgroundremover/default.nix:5]: Blank line between the parameters and the function call

File: pkgs/development/python-modules/backgroundremover/default.nix [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  pname = "backgroundremover";
+  name = pname;
```
Comment (lucasew) [inline pkgs/development/python-modules/backgroundremover/default.nix:7]: remove name, use pname only

name is calculated as "${pname}-${version}" automatically

File: pkgs/development/python-modules/backgroundremover/default.nix [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  version = "0.2.1";
+
+  meta = with lib; {
+    description =
+      "BackgroundRemover is a command line tool to remove background from image and video, made by nadermx to power https://BackgroundRemoverAI.com";
+    platforms = platforms.all;
+    homepage = "https://github.com/nadermx/backgroundremover";
+    downloadPage = "https://github.com/nadermx/backgroundremover/releases";
+    license = licenses.mit;
+    maintainers = with maintainers; [ jayrovacsek ];
```
Comment (lucasew) [inline pkgs/development/python-modules/backgroundremover/default.nix:17]: `maintainers = [ maintainers.jayrovacsek ];`

File: pkgs/development/python-modules/backgroundremover/default.nix [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    requests
+    scikitimage
+    scipy
+    six
+    torch
+    torchvision
+    tqdm
+    urllib3
+    waitress
+    gdown
```
Comment (lucasew) [inline pkgs/development/python-modules/backgroundremover/default.nix:43]: sort the list

File: pkgs/development/python-modules/commandlines/default.nix [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  version = "0.4.1";
+
+  meta = with lib; {
+    description =
+      "Commandlines is a Python library for command line application development that supports command line argument parsing";
+    platforms = platforms.all;
+    homepage = "https://github.com/chrissimpkins/commandlines";
+    downloadPage = "https://github.com/chrissimpkins/commandlines/releases";
+    license = licenses.mit;
+    maintainers = with maintainers; [ jayrovacsek ];
```
Comment (lucasew) [inline pkgs/development/python-modules/commandlines/default.nix:14]: same maintainer thing as the other file

File: pkgs/development/python-modules/commandlines/default.nix [hunk 7]
```diff
@@ -0,0 +1,21 @@
+{ buildPythonPackage, fetchPypi, lib }:
+buildPythonPackage rec {
+  pname = "commandlines";
+  name = pname;
```
Comment (lucasew) [inline pkgs/development/python-modules/commandlines/default.nix:4]: remove `name`, same reason as the other file

File: pkgs/development/python-modules/commandlines/default.nix [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  version = "0.4.1";
+
+  meta = with lib; {
```
Comment (lucasew) [inline pkgs/development/python-modules/commandlines/default.nix:7]: meta is the last thing

File: pkgs/development/python-modules/commandlines/default.nix [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    description =
+      "Commandlines is a Python library for command line application development that supports command line argument parsing";
+    platforms = platforms.all;
+    homepage = "https://github.com/chrissimpkins/commandlines";
+    downloadPage = "https://github.com/chrissimpkins/commandlines/releases";
+    license = licenses.mit;
+    maintainers = with maintainers; [ jayrovacsek ];
+  };
+
+  src = fetchPypi {
```
Comment (lucasew) [inline pkgs/development/python-modules/commandlines/default.nix:17]: src right below version

File: pkgs/development/python-modules/hsh/default.nix [hunk 10]
```diff
@@ -0,0 +1,23 @@
+{ buildPythonPackage, fetchPypi, lib, commandlines }:
+buildPythonPackage rec {
+  pname = "hsh";
+  name = pname;
+  version = "1.1.0";
+
+  meta = with lib; {
```
Comment (lucasew) [inline pkgs/development/python-modules/hsh/default.nix:7]: meta is the last thing

File: pkgs/development/python-modules/hsh/default.nix [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    description =
+      "hsh is a cross-platform command line application that generates file hash digests and performs file integrity checks via file hash digest comparisons.";
+    platforms = platforms.all;
+    homepage = "https://github.com/chrissimpkins/hsh";
+    downloadPage = "https://github.com/chrissimpkins/hsh/releases";
+    license = licenses.mit;
+    maintainers = with maintainers; [ jayrovacsek ];
+  };
+
+  src = fetchPypi {
```
Comment (lucasew) [inline pkgs/development/python-modules/hsh/default.nix:17]: src right after version

File: pkgs/development/python-modules/hsh/default.nix [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    inherit pname version;
+    sha256 = "sha256-wE5DrFOP6vsCnbo8SXIgenBPX83w7icevd3QPZa134U=";
+  };
+
+  propagatedBuildInputs = [ commandlines ];
```
Comment (lucasew) [inline pkgs/development/python-modules/hsh/default.nix:22]: this right after src

File: pkgs/development/python-modules/pymatting/default.nix [hunk 13]
```diff
@@ -0,0 +1,26 @@
+{ buildPythonPackage, fetchPypi, lib, numpy
+, pillow, numba, scipy }:
+buildPythonPackage rec {
+  pname = "PyMatting";
+  name = pname;
```
Comment (lucasew) [inline pkgs/development/python-modules/pymatting/default.nix:5]: remove name

File: pkgs/development/python-modules/pymatting/default.nix [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+  version = "1.1.8";
+
+  meta = with lib; {
```
Comment (lucasew) [inline pkgs/development/python-modules/pymatting/default.nix:8]: meta at the end

File: pkgs/development/python-modules/pymatting/default.nix [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+    description = "PyMatting: A Python Library for Alpha Matting";
+    platforms = platforms.all;
+    homepage = "https://github.com/pymatting/pymatting";
+    downloadPage = "https://github.com/pymatting/pymatting/releases";
+    license = licenses.mit;
+    maintainers = with maintainers; [ jayrovacsek ];
+  };
+
+  src = fetchPypi {
```
Comment (lucasew) [inline pkgs/development/python-modules/pymatting/default.nix:17]: src after version

File: pkgs/development/python-modules/pymatting/default.nix [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+    inherit pname version;
+    sha256 = "sha256-pzUI7wh0mWgx39KE47xjFRoJSE14n9EO3AP0mzFTKsw=";
+  };
+
+  doCheck = false;
+
+  propagatedBuildInputs = [ numpy pillow numba scipy ];
```
Comment (lucasew) [inline pkgs/development/python-modules/pymatting/default.nix:24]: this after src

File: pkgs/development/python-modules/pymatting/default.nix [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+    license = licenses.mit;
+    maintainers = with maintainers; [ jayrovacsek ];
+  };
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "sha256-pzUI7wh0mWgx39KE47xjFRoJSE14n9EO3AP0mzFTKsw=";
+  };
+
+  doCheck = false;
```
Comment (lucasew) [inline pkgs/development/python-modules/pymatting/default.nix:22]: check stuff is after nativeBuildInputs

**Review by lucasew — CHANGES_REQUESTED** [review]
some standard stuff

change the name of this pr to `backgroundremover: init at 0.2.1`

and after that put each change in it's own commit (I start with `git reset --soft HEAD~1` then I start building the commits)

**Comment by lucasew** [comment]
BTW pro-tip: it's more convenient to run nixpkgs-review with the `--post-result` flag instead of pasting the output in the PR description.

I use that a lot to pipeline the reviews here, sometimes I can't actually review, but I can test if it builds.

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 231295` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>python310Packages.backgroundremover</li>
    <li>python310Packages.backgroundremover.dist</li>
    <li>python310Packages.commandlines</li>
    <li>python310Packages.commandlines.dist</li>
    <li>python310Packages.hsh</li>
    <li>python310Packages.hsh.dist</li>
    <li>python310Packages.pymatting</li>
    <li>python310Packages.pymatting.dist</li>
    <li>python311Packages.commandlines</li>
    <li>python311Packages.commandlines.dist</li>
    <li>python311Packages.hsh</li>
    <li>python311Packages.hsh.dist</li>
  </ul>
</details>

**Comment by alyssais** [comment]
@lucasew I think this level of style nitpicking is excessive.  We don't have that level of consistent style across the project, and even the [style rules we do have in the manual](https://nixos.org/manual/nixpkgs/unstable/#sec-syntax) are not widely followed.  Because we don't have a consistent style, being too strict about style can be very offputting for contributors, because one reviewer might tell them style should be one way, and another will tell them something different, which is very arbitrary from the contributor's point of view.

**Comment by lucasew** [comment]
I agree with you. Some stuff is kinda C++y where people enforce different patterns that sometimes even conflict with each other. I think we need to work towards a written document or something like that with the consensus of the whole community. With that some kind of linter can be used to automate this part. A nitbot xD.

**Comment by JayRovacsek** [comment]
Extremely big thank you to all of the constructive feedback - I'll address all comments and update this pull request as soon as I can.

Apologies for the singular commit for multiple package additions - this was a misunderstanding on my part from feedback within the unofficial Discord server but I am more than happy to resolve this issue which might alleviate what I suspect the issue @AndersonTorres is trying to raise (plus feedback [here](https://github.com/NixOS/nixpkgs/pull/224579) which likely marries up to the same issue of a singular commit)

I'll on-board all the recommended  style/format changes you've made @lucasew - I'm extremely keen on keeping consistency with the current style and had naively applied a simple `nixfmt` and `deadnix` run across the change-set here.

Again, thanks all :heart: :+1:

**Comment by JayRovacsek** [comment]
Hi all, I recognise the above problems can be resolved via some git-fu, I however lack the skills to achieve this and have placed my fork of nixpkgs into a bad state to create more pull requests. I am going to close this request, but will reopen a request with all of the above changes applied and commit messages correct.
I'll link this request in that request and vice-versa once it is created shortly.

Again, thanks all for the feedback :heart:

**Comment by JayRovacsek** [comment]
As described in last comment, please find the new pull request here: https://github.com/NixOS/nixpkgs/pull/231950

**Comment by JayRovacsek** [comment]
Result of `nixpkgs-review pr 231295` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>python310Packages.backgroundremover</li>
    <li>python310Packages.backgroundremover.dist</li>
    <li>python310Packages.commandlines</li>
    <li>python310Packages.commandlines.dist</li>
    <li>python310Packages.hsh</li>
    <li>python310Packages.hsh.dist</li>
    <li>python310Packages.pymatting</li>
    <li>python310Packages.pymatting.dist</li>
    <li>python311Packages.commandlines</li>
    <li>python311Packages.commandlines.dist</li>
    <li>python311Packages.hsh</li>
    <li>python311Packages.hsh.dist</li>
  </ul>
</details>

**Pull request closed without merging**